MiniApp/Classes/ui/MiniAppSharePreviewViewController.swift

```diff
@@ -0,0 +1,107 @@
+
+    let promotionText: String
+    let promotionImageUrl: String
+
+    lazy var promotionLabel: UILabel = {
+        let view = UILabel()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.text = promotionText
+        view.textAlignment = .center
+        view.numberOfLines = 0
+        return view
+    }()
+
+    lazy var promotionImageView: UIImageView = {
+        let view = UIImageView()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.backgroundColor = .secondarySystemBackground
+        view.layer.cornerRadius = 10
+        view.clipsToBounds = true
+        return view
+    }()
+
+    lazy var loadingIndicatorView: UIActivityIndicatorView = {
+        let view = UIActivityIndicatorView()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        return view
+    }()
+
+    lazy var closeBarButton: UIBarButtonItem = {
+        let view = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(closePressed))
+        return view
+    }()
+
+    init(promotionText: String, promotionImageUrl: String) {
+        self.promotionText = promotionText
+        self.promotionImageUrl = promotionImageUrl
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder: NSCoder) { return nil }
+
+    public override func viewDidLoad() {
+        super.viewDidLoad()
+
+        title = "Share"
+        view.backgroundColor = .systemBackground
+        navigationItem.setRightBarButton(closeBarButton, animated: false)
+
+        setupUI()
+    }
+
+    public override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        loadPreviewImage()
+    }
+
+    func setupUI() {
+        view.addSubview(promotionLabel)
+        NSLayoutConstraint.activate([
+            promotionLabel.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 20),
+            promotionLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor),
+            promotionLabel.trailingAnchor.constraint(equalTo: view.trailingAnchor)
+        ])
+
+        view.addSubview(promotionImageView)
+        NSLayoutConstraint.activate([
+            promotionImageView.topAnchor.constraint(equalTo: promotionLabel.bottomAnchor, constant: 20),
+            promotionImageView.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 20),
+            promotionImageView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -20),
+            promotionImageView.heightAnchor.constraint(equalTo: view.widthAnchor, multiplier: 0.667, constant: -40)
+        ])
+
+        promotionImageView.addSubview(loadingIndicatorView)
+        NSLayoutConstraint.activate([
+            loadingIndicatorView.centerXAnchor.constraint(equalTo: promotionImageView.centerXAnchor),
+            loadingIndicatorView.centerYAnchor.constraint(equalTo: promotionImageView.centerYAnchor)
+        ])
```

Comment:
All of this (and all the lazy variables declarations) could be simplified by the use of storyboards 😅

Reply:
true but I prefer coded ui 😄  if required I can change it

Reply:
that's fine, one day we'll migrate to SwiftUI so we kind of have a taste of it

Reply:
yeah best of both comes together then 🥇